src/software/ai/hl/stp/tactic/attacker/attacker_tactic.h

```diff
@@ -1,34 +1,125 @@
+     * of its current Skill.
+     *
+     * IMPORTANT: updatePrimitive should not be called on this tactic if it is
+     * suspended. This is because a suspended tactic will not yield any primitives
+     * to execute.
+     *
+     * @return true if the tactic has suspended execution, false otherwise
+     */
+    bool suspended() const;
+
+    /**
+     * If the AttackerTactic has temporarily suspended execution of its current Skill,
+     * calling this method will update the AttackerTactic with the given world and try
+     * to resume execution (i.e. leave the suspended state).
+     *
+     * If the AttackerTactic is not suspended, then calling this method effectively
+     * does nothing.
+     *
+     * @param world_ptr the world pointer
+     *
+     * @return true if the tactic is still suspended after trying to resume execution,
+     * false otherwise
+     */
+    bool tryResumingIfSuspended(const WorldPtr& world_ptr);
 
     /**
-     * Evaluate the AttackerTactic and start a new iteration
+     * Terminate the current episode of the AttackerTactic and reset the tactic
+     * for a new episode
      *
-     * @param score score between [-1, 1] rating the success of the AttackerTactic
-     * for the current iteration
+     * @param world_ptr the World at the end of the current episode
      */
-    void evaluate(double score);
+    void terminate(const WorldPtr& world_ptr);
+
+    /**
+     * Gets a SkillState containing details about the current state of the skill
+     * that the AttackerTactic is executing.
+     * 
+     * @return the current state of the Skill that the tactic is executing
+     */
+    SkillState getSkillState() const;
+
+    /**
+     * Log visualize the SkillState for this AttackerTactic.
+     *
+     * @param world the current state of the world
+     */
+    void visualizeSkillState(const World& world);
 
    private:
-    std::shared_ptr<Strategy> strategy;
+    // The shared Strategy used by all of AI
+    std::shared_ptr<Strategy> strategy_;
+
+    // The Q-function of the attacker agent's policy
+    std::shared_ptr<LinearQFunction<AttackerMdpState, AttackerMdpAction>> q_function_;
+
+    // The action selection strategy of attacker agent's policy
+    std::shared_ptr<SoftmaxStrategy<AttackerMdpState, AttackerMdpAction>>
+        action_selection_strategy_;
 
-    SkillGraph skill_graph_;
-    std::shared_ptr<Skill> current_skill_;
+    // The policy that the attacker agent will follow
+    QPolicy<AttackerMdpState, AttackerMdpAction> policy_;
+
+    // The reward function for the attacker agent
+    AttackerMdpRewardFunction attacker_mdp_reward_function_;
+
+    // The current skill that the attacker is executing
+    std::unique_ptr<Skill> current_skill_;
+
+    // File to load Q-function weights from for initialization
+    inline static const std::string ATTACKER_MDP_Q_FUNCTION_INITIAL_WEIGHTS_FILE =
+        "software/ai/evaluation/q_learning/attacker_mdp_q_function_weights.csv";
```

Comment:
we may or may not need to add this to bazel as a data dependency because of the way bazel sandboxes builds

Reply:
I think I did this already
https://github.com/UBC-Thunderbots/Software/blob/c235594764cc6eee017d7c9da4a7717f5d2c7632/src/software/ai/evaluation/q_learning/BUILD#L3

https://github.com/UBC-Thunderbots/Software/blob/c235594764cc6eee017d7c9da4a7717f5d2c7632/src/software/ai/hl/stp/tactic/attacker/BUILD#L7-L9